src/storage/v2/durability/snapshot.cpp

```diff
@@ -157,17 +187,378 @@ SnapshotInfo ReadSnapshotInfo(const std::filesystem::path &path) {
   return info;
 }
 
-RecoveredSnapshot LoadSnapshot(const std::filesystem::path &path, utils::SkipList<Vertex> *vertices,
-                               utils::SkipList<Edge> *edges,
-                               std::deque<std::pair<std::string, uint64_t>> *epoch_history,
-                               NameIdMapper *name_id_mapper, std::atomic<uint64_t> *edge_count, Config::Items items) {
+std::vector<BatchInfo> ReadBatchInfos(Decoder &snapshot) {
+  std::vector<BatchInfo> infos;
+  const auto infos_size = snapshot.ReadUint();
+  if (!infos_size.has_value()) {
+    throw RecoveryFailure("Invalid snapshot data!");
+  }
+  infos.reserve(*infos_size);
+
+  for (auto i{0U}; i < *infos_size; ++i) {
+    const auto offset = snapshot.ReadUint();
+    if (!offset.has_value()) {
+      throw RecoveryFailure("Invalid snapshot data!");
+    }
+
+    const auto count = snapshot.ReadUint();
+    if (!count.has_value()) {
+      throw RecoveryFailure("Invalid snapshot data!");
+    }
+    infos.push_back(BatchInfo{*offset, *count});
+  }
+  return infos;
+}
+
+template <typename TFunc>
+void LoadPartialEdges(const std::filesystem::path &path, utils::SkipList<Edge> &edges, const uint64_t from_offset,
+                      const uint64_t edges_count, const Config::Items items, TFunc get_property_from_id) {
+  Decoder snapshot;
+  snapshot.Initialize(path, kSnapshotMagic);
+
+  // Recover edges.
+  auto edge_acc = edges.access();
+  uint64_t last_edge_gid = 0;
+  spdlog::info("Recovering {} edges.", edges_count);
+  if (!snapshot.SetPosition(from_offset)) throw RecoveryFailure("Couldn't read data from snapshot!");
+  for (uint64_t i = 0; i < edges_count; ++i) {
+    {
+      const auto marker = snapshot.ReadMarker();
+      if (!marker || *marker != Marker::SECTION_EDGE) throw RecoveryFailure("Invalid snapshot data!");
+    }
+
+    if (items.properties_on_edges) {
+      // Insert edge.
+      auto gid = snapshot.ReadUint();
+      if (!gid) throw RecoveryFailure("Invalid snapshot data!");
+      if (i > 0 && *gid <= last_edge_gid) throw RecoveryFailure("Invalid snapshot data!");
+      last_edge_gid = *gid;
+      spdlog::debug("Recovering edge {} with properties.", *gid);
+      auto [it, inserted] = edge_acc.insert(Edge{Gid::FromUint(*gid), nullptr});
+      if (!inserted) throw RecoveryFailure("The edge must be inserted here!");
+
+      // Recover properties.
+      {
+        auto props_size = snapshot.ReadUint();
+        if (!props_size) throw RecoveryFailure("Invalid snapshot data!");
+        auto &props = it->properties;
+        for (uint64_t j = 0; j < *props_size; ++j) {
+          auto key = snapshot.ReadUint();
+          if (!key) throw RecoveryFailure("Invalid snapshot data!");
+          auto value = snapshot.ReadPropertyValue();
+          if (!value) throw RecoveryFailure("Invalid snapshot data!");
+          props.SetProperty(get_property_from_id(*key), *value);
```

Comment:
Does it make sense to use `InitProperties` method here as it can batch process the properties?

Reply:
I didn't think about the part in which properties are stored in order, so properties are added to the end of the buffer. The only thing is if there are a lot of properties, the buffer will be needed to be copied a few times, with a new buffer being allocated.

One more thing that happens with calling SetProperty a few times is that each time we need to do FindSpecificPropertyAndBufferInfo. This iterates over the whole PropertyStore. You can check that even on a few properties I think we gained enough with `InitProperties`

I think we can still gain here in the following way:

1. we initialize std::vector<PropertyId, PropertyValue> properties(props_size); This way we don't have to deal with std::map because properties are already ordered.
2. update InitProperties to accept also std::vector
3. In that case PropertyStore will need to do only one allocation of buffer and it should be pretty fast, faster than building std::map or doing n * SetProperty due to resizing of buffer and copying of buffer. I understand it is fast, but it happened that SetProperty was actually problematic on CreateLocalVertex, and only in a few cases can all properties fit inside 15 bytes. If you have two doubles, that is already over the limit if I am not wrong.


If this makes sense to you I would go in this direction, otherwise, we can leave it as it is.